Update SDK-PR-guide.md to include branch lockdown label

marcpopMSFT opened the pull request:
I included a comment in the PR guide about the lockdown label
Additionally, I added a workflow to automatically add that label.

- The branch is generally open between the first tuesday of the month when branding goes in until the third tuesday of the month (basically a week after the internal to public merge).
- Outside of that window, let's automatically add the label to any 8/9 base branch PR
- The workflow uses the permissions of the target branch as well as the source for the target branch so won't run until it's merged
- I configured it to run on all PRs but will only check into main. 
- When it runs, it'll search all PRs and label them all so we don't need to configure it to run elsewhere.
- I check if it's a PR against the 8 or 9 branches filtering out 9.0.3xx
- I check if it already has the label

marcpopMSFT committed:
Update SDK-PR-guide.md

joeloff commented on documentation/project-docs/SDK-PR-guide.md:
```diff
@@ -26,7 +26,7 @@ After the minor release locks down, it transitions into servicing / "QB" (Quarte
 The .NET SDK has monthly servicing releases (e.g.`7.0.100` VS `7.0.101`) aligning with the .NET Runtime servicing releases. These are for top fixes and security updates only to limit risk.
 Any servicing release is open for check-ins from the day the [branding PRs](https://github.com/dotnet/sdk/pulls?q=is%3Apr+branding) are merged (~1st of each month) and when code complete is (typically two weeks later).
-The servicing branches are locked from the time of code complete to the next branding in case we need to respin any monthly release. Final signoffs are typically in the last week of each month.
+The servicing branches are locked from the time of code complete to the next branding in case we need to respin any monthly release. We use the `Branch Lockdown` label to mark PRs that should go into the next servicing release but the branch is currently not open. Final signoffs are typically in the last week of each month.
```
Is there a way for us to have some automated action can calculate lockdown dates and have the labels automatically apply when a PR is created? The dates usually vary based on when the release date is, etc. so I'm guessing it's not easy, but perhaps it can be scoped to really critical periods

nagilson replied:
That would be pretty awesome.

marcpopMSFT replied:
I added a workflow yml that should add the label to PRs by maestro to the release/8 and release/9 branches (excluding release/9.0.3xx) as long as it's after the third Tuesday of the month or before the first Tuesday of the month (when branding happens). I don't know how to test this though.

marcpopMSFT committed:
Create add-lockdown-label.yml

marcpopMSFT committed:
add a PR trigger so I can test this

marcpopMSFT committed:
Fix date calculation for first and third Tuesdays

marcpopMSFT committed:
Add debug logs and filter PRs

marcpopMSFT committed:
Add debug logs for PR filtering process

marcpopMSFT committed:
Fix PR filtering by baseRefName instead of headRefName

marcpopMSFT committed:
Update title and permissions

marcpopMSFT committed:
Increase PR list query limit to 200

marcpopMSFT committed:
Update PR filtering and labeling logic

jeffhandley commented on .github/workflows/add-lockdown-label.yml:
```diff
@@ -0,0 +1,81 @@
+on:
+  schedule:
+    # Run daily at 00:00 UTC
+    - cron: '0 0 * * *'
```
Nit for what it's worth: GitHub recommends picking a time other than midnight UTC and even suggests picking arbitrary minutes within the hour to avoid resource contention at the most popular time for scheduled workflows to kick off.

jeffhandley commented on .github/workflows/add-lockdown-label.yml:
```diff
@@ -0,0 +1,81 @@
+    # Run daily at 00:00 UTC
+    - cron: '0 0 * * *'
+  workflow_dispatch: # Allows manual triggering of the workflow
+  pull_request:
+    branches:
+      - main
```
This is where your permissions issue exists. You need to use the `pull_request_target` event instead of `pull_request`.

* `pull_request` runs in the context of the pull request's content (and would actually execute the workflow as defined within the pull request if the workflow file is updated in the pull request). Because of that context, the GitHub Token provided to the workflow in that context in read-only when the pull request comes from a fork.
* `pull_request_target` runs in the context of the pull request's base/target branch's content. It will therefore run the workflow definition as already committed to that branch (in your case, `main`), and the GitHub Token is therefore given the permissions you need for this type of workflow.

Here are some references:

[`pull_request` Workflows in forked repositories](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#workflows-in-forked-repositories)

> The GITHUB_TOKEN has read-only permissions in pull requests from forked repositories.
 
[`pull_request_target`](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request_target)

> This event runs in the context of the base of the pull request, rather than in the context of the merge commit, as the pull_request event does. This prevents execution of unsafe code from the head of the pull request that could alter your repository or steal any secrets you use in your workflow. **This event allows your workflow to do things like label or comment on pull requests from forks**. Avoid using this event if you need to build or run code from the pull request.

marcpopMSFT replied:
@jeffhandley that makes sense. I'll try changing this and retriggering and see if that works.

marcpopMSFT replied:
When I changed that, it didn't trigger the workflow presumably because the target branch doesn't have this. So I don't think I can actually test this in the PR so I'll need to merge first and then test.

marcpopMSFT committed:
Update cron schedule and permissions in workflow

marcpopMSFT committed:
Remove pull_request_target and pull-requests permissions

MiYanni commented on .github/workflows/add-lockdown-label.yml:
```diff
@@ -0,0 +1,78 @@
+          if [[ "$current_date" > "$third_tuesday" || "$current_date" < "$first_tuesday" ]]; then
+            echo "Within the label period. Adding labels to PRs..."
+
+            # Get all open PRs created by app/dotnet-maestro and targeting branches release/8* and release/9* excluding release/9.0.3xx
```
So, our branch-to-branch codeflow PRs use `app/github-actions`. Also, why are we limiting the PRs based on who created them? If the branch is locked down, *all PRs* targeting those branches should have "Branch Lockdown", right?

marcpopMSFT replied:
At least today, I use branch lockdown also when the branch unlocks to find all PRs that should go in. If I don't limit this, I'll pick up other PRs that just happened to target those branches. Not a good excuse and I think your argument to just tag all of them is probably stronger.

MiYanni commented on .github/workflows/add-lockdown-label.yml:
```diff
@@ -0,0 +1,78 @@
+            echo "PRs by app/dotnet-maestro JSON: $prs_by_author"
+            echo "Count of PRs by app/dotnet-maestro: $(echo "$prs_by_author" | jq length)"
+
+            echo "Filtering PRs targeting release/8* and release/9* excluding release/9.0.3xx..."
+            prs_targeting_branches=$(echo "$prs_by_author" | jq '[.[] | select(.baseRefName | test("^release/[89].*")) | select(.baseRefName | test("^release/9.0.3xx") | not)]')
```
This will change (somewhat) regularly as we create new branches. If we need a branch filter, it should just be a separate file (or variable/parameter) to this action. But, I don't know GitHub Actions to know what makes the most sense here.

marcpopMSFT replied:
I'd worry more about a variable if we were using this in other places. We'd need to change it wherever it lives so I don't think it matters if it's in this file or elsewhere.

MiYanni commented on .github/workflows/add-lockdown-label.yml:
```diff
@@ -0,0 +1,78 @@
+      - name: Checkout repository
+        uses: actions/checkout@v2
+
+      - name: Install jq
+        run: sudo apt-get install -y jq
```
What it jq? It looks to be some kind of filtering mechanism (like what is native to PowerShell) but as a separate Bash module.

marcpopMSFT replied:
It's a json parsing tool that copilot recommended.

jeffhandley replied:
[jq](https://jqlang.org/) is very commonly used in GitHub workflows. Great command-line tool for processing JSON data.

MiYanni replied:
I'm not *trying* to promote it, but PowerShell can natively parse JSON. Apparently, you can just [set the shell](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell) for use in these steps.

MiYanni commented on .github/workflows/add-lockdown-label.yml:
```diff
@@ -0,0 +1,78 @@
+
+on:
+  schedule:
+    # Run daily at 00:00 UTC
+    - cron: '0 23 * * *'
```
I'm not convinced this is the right approach overall. The reason is, this runs once a day. Standard label bots run as **PRs are created**. That's what triggers them to run. Sure, this process will work on pre-existing PRs just fine. But as things attempt to "flow" from other repos during the day, domestic cat might accidentally merge things they shouldn't.

Also, triggering this at the "start of the day" like this, will there be any issues with the date calculation below? Most processes like this avoid the exact start of a day because stuff like daylight savings time exists. Plus, this runs right in the middle of Jason's shift, I believe.

marcpopMSFT replied:
So you think it should trigger on every PR do these branches then? That means it runs a lot of times but maybe that's ok. I wonder then if it runs on PRs, if we could just run it against the PR that's being created rather than querying all the PRs.

MiYanni replied:
I'm fairly certain you can trigger on PR creation. But I don't know GitHub Actions.

edvilme replied:
Yes, you can. I believe you can also trigger it on other PR events (opened, closed, assigned, etc...)
https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request

marcpopMSFT committed:
Modify the logic to run on the PR creation and apply to all PRs to those branches

marcpopMSFT commented:
/backport to release/9.0.2xx

marcpopMSFT committed:
Update branch lockdown label workflow

marcpopMSFT committed:
Improve branch matching and logging in workflow

marcpopMSFT committed:
Change event trigger to pull_request_target

marcpopMSFT committed:
Remove jq installation from workflow

marcpopMSFT committed:
Update workflow to add jq installation and PR filtering

The prior version of this would only label new PRs not PRs created before the date. This searches all PRs that match the requirements whenever a new PR is created targeting these branches.

marcpopMSFT committed:
Remove debug echo statements from workflow

marcpopMSFT committed:
Update workflow to use pull_request_target

marcpopMSFT committed:
Update workflow to add permissions and increase PR limit

marcpopMSFT committed:
Remove author filter from PR query
